src/Agent.Worker/StepsRunner.cs

```diff
@@ -125,20 +131,34 @@ public sealed class StepsRunner : AgentService, IStepsRunner
             }
         }
 
-        private async Task RunStepAsync(IExecutionContext jobContext, IStep step)
+        private async Task RunStepAsync(IExecutionContext jobContext, IStep step, bool alreadyCanceled, DT.INode conditionTree)
         {
-            // Run the step.
+            // Start the step.
             Trace.Info("Starting the step.");
             step.ExecutionContext.Section(StringUtil.Loc("StepStarting", step.DisplayName));
-            if (step.Timeout != null)
+            step.ExecutionContext.SetTimeout(timeout: step.Timeout);
+            Task stepTask = step.RunAsync();
+            try
+            {
+                // When alreadyCanceled is true, it means the job cancellation token was already signaled
+                // before the condition was evaluated.
+                Task.WaitAny(new[] { stepTask }, alreadyCanceled ? CancellationToken.None : jobContext.CancellationToken);
+            }
+            catch (OperationCanceledException)
             {
-                step.ExecutionContext.SetTimeout(timeout: step.Timeout.Value);
+                // Test the condition again. The job was canceled after the condition was originally evaluated.
+                var expressionManager = HostContext.GetService<IExpressionManager>();
+                if (!expressionManager.Evaluate(jobContext, step.ExecutionContext, conditionTree, hostTracingOnly: true))
```

Comment:
evaluation could fail here too, so this should be in a try/catch and also treated as critical failure.